### PR TITLE
Add Ability to Move and Re-order Columns in "Game List" Mode

### PR DESCRIPTION
**Pull request by RetraCarteR**
As someone who primarily uses the "Game List" mode in RMG's ROM browser, I feel like it could use some improvement. While perfectly serviceable, it's pretty barebones compared to the ROM browsers in other emulators that use Qt6. It's seriously lacking in customization options beyond the ability to resize columns and change how ROMs are sorted, and I've decided to take it upon myself to rectify that (with your approval, of course).

I came up with a three-step plan to bring RMG's "Game List" mode up to snuff, and implementing the ability for users to rearrange the columns was step one. Step two will be to implement the ability to hide and un-hide columns via a context menu that appears when users right-click the column header area, and step three will be to add more columns so that users can view more advanced information about each ROM (such as its file extension, file size, and region) without needing to go into the "ROM Information" sub-menu for each game.

I understand that you may have reservations about adding more columns, since that would clutter up the interface, but that's where steps one and two come in. Any hypothetical newly-added columns would be hidden by default, but users would have the ability to un-hide them if they wish. Or hide any of the default columns if they want to. Or re-order them from left-to-right however they'd like. The menu will still look the same by default, but users would have the ability to customize it pretty much however they want. I can't think of any good reason not to at least give them the option to do so.

I've included more information about how I've implemented step 1 in the description of the included commit. I plan on implementing steps two and three at some point in the future, but I currently have no estimation for how long that will take. I'm fully prepared to implement these changes on my own without any outside help, so you won't have to do anything aside from merging my commits as I make them. Of course, if you don't want to change the ROM browser at all, that's fine too. Just say the word and I'll close this PR, abandon this plan, and move on with my life, no hard feelings.

Thank you for all of the work you've done and are continuing to do on RMG. I already think it's excellent, and I'd be honored if you gave me the go-ahead to improve it even more.

### CODE REVIEW

**Comment by Rosalie241** [comment]
> As someone who primarily uses the "Game List" mode in RMG's ROM browser, I feel like it could use some improvement. While perfectly serviceable, it's pretty barebones compared to the ROM browsers in other emulators that use Qt6. It's seriously lacking in customization options beyond the ability to resize columns and change how ROMs are sorted, and I've decided to take it upon myself to rectify that (with your approval, of course).
> 

That's true, some ROM Browser things (like making the columns customizable) are still TODO, so I haven't gotten around to it yet!

> I came up with a three-step plan to bring RMG's "Game List" mode up to snuff, and implementing the ability for users to rearrange the columns was step one. Step two will be to implement the ability to hide and un-hide columns via a context menu that appears when users right-click the column header area, and step three will be to add more columns so that users can view more advanced information about each ROM (such as its file extension, file size, and region) without needing to go into the "ROM Information" sub-menu for each game.
> 

Sounds fine to me.

> I understand that you may have reservations about adding more columns, since that would clutter up the interface, but that's where steps one and two come in. Any hypothetical newly-added columns would be hidden by default, but users would have the ability to un-hide them if they wish. Or hide any of the default columns if they want to. Or re-order them from left-to-right however they'd like. The menu will still look the same by default, but users would have the ability to customize it pretty much however they want. I can't think of any good reason not to at least give them the option to do so.
>

Don't worry, RMG was meant to be customizable and the kind of stuff you mentioned has been on my long-term TODO, I just didn't have the energy because I'm constantly working on other stuff in RMG.

> I've included more information about how I've implemented step 1 in the description of the included commit. I plan on implementing steps two and three at some point in the future, but I currently have no estimation for how long that will take. I'm fully prepared to implement these changes on my own without any outside help, so you won't have to do anything aside from merging my commits as I make them. Of course, if you don't want to change the ROM browser at all, that's fine too. Just say the word and I'll close this PR, abandon this plan, and move on with my life, no hard feelings.
> 

Thank you for wanting to help out and don't worry about the time-frame for the other stuff, there's no rush.

The commit looks fine from a first glance, will review it shortly!

> Thank you for all of the work you've done and are continuing to do on RMG. I already think it's excellent, and I'd be honored if you gave me the go-ahead to improve it even more.

I'm glad you enjoy using RMG, I try my best to give a good user experience and I'm glad someone wants to help out :heart:

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 1]
```diff
@@ -616,12 +630,19 @@ void RomBrowserWidget::on_RomBrowserThread_Finished(bool canceled)
     else
     {
         // use settings' width
-        for (int i = 0; i < (columnSizes.size() - 1); i++)
+        for (int i = 0; i < columnSizes.size(); i++)
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:633]: Why did you change this if I may ask? the last column should always be stretched, otherwise the width of the last column looks weird, no?

Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:633]: Keeping the last column stretched causes weird issues when moving the columns around. A lot of the time, one of the first two columns stays stretched instead of the last one depending on how they're arranged when the layout is saved. It's pretty late in my time zone right now and I'm about to go to bed, but I'll try and see if I can figure out a better solution for that once I wake up.

**Comment by RetraCarteR** [comment]
All right, I think I've finally gotten the re-ordering feature to work while keeping the last column stretched. The trick was to wait to call "setStretchLastSection" until after the column sizes have been set on startup. You can look at the new commit I posted for more details.

**Comment by Rosalie241** [comment]
Sadly it seems the `this->listViewWidget->horizontalHeader()->setStretchLastSection(true);` causes crashes for me when refreshing the ROM browser, after some trial & error, I'm unable to fix it so hmm.

**Comment by RetraCarteR** [comment]
Hang on, let me test real quick.

**Comment by RetraCarteR** [comment]
It's crashing for me too. I can't believe I didn't think to try refreshing the ROM list. Working on fixing it now.

**Comment by RetraCarteR** [comment]
And fixed! Turns out that my changes to the initialization code depended on "setStretchLastSection" being false at first, so reinitializing by refreshing the ROM list makes it fail because it's still set to "true" from the first time it ran the code. I fixed it by temporarily setting "setStretchLastSection" to false at the start of initialization.

**Comment by Rosalie241** [comment]
Thank you :heart:
